sql-plugin/src/main/scala/com/nvidia/spark/rapids/TableCompressionCodec.scala

```diff
@@ -0,0 +1,367 @@
+      inputLength: Long): Unit
+
+  /**
+   * Create a batched compressor instance
+   * @param maxBatchMemorySize The upper limit in bytes of temporary and output memory usage at
+   *                           which a batch should be compressed. A single table that requires
+   *                           temporary and output memory above this limit is allowed but will
+   *                           be compressed individually.
+   * @return batched compressor instance
+   */
+  def createBatchCompressor(maxBatchMemorySize: Long): BatchedTableCompressor
+
+  /**
+   * Create a batched compressor instance
+   * @param maxBatchMemorySize The upper limit in bytes of temporary and output memory usage at
+   *                           which a batch should be decompressed. A single buffer that requires
+   *                           temporary and output memory above this limit is allowed but will
+   *                           be decompressed individually.
+   * @return batched decompressor instance
+   */
+  def createBatchDecompressor(maxBatchMemorySize: Long): BatchedBufferDecompressor
+}
+
+object TableCompressionCodec {
+  private val codecNameToId = Map(
+    "copy" -> CodecType.COPY)
+
+  /** Get a compression codec by short name or fully qualified class name */
+  def getCodec(name: String): TableCompressionCodec = {
+    val codecId = codecNameToId.getOrElse(name.toLowerCase(Locale.ROOT),
+      throw new IllegalArgumentException(s"Unknown table codec: $name"))
+    getCodec(codecId)
+  }
+
+  /** Get a compression codec by ID, using a cache. */
+  def getCodec(codecType: Byte): TableCompressionCodec = {
+    codecType match {
+      case CodecType.COPY => new CopyCompressionCodec
+      case _ => throw new IllegalArgumentException(s"Unknown codec ID")
+    }
+  }
+}
+
+/**
+ * Base class for batched compressors
+ * @param maxBatchMemorySize The upper limit in bytes of temporary and output memory usage at
+ *                           which a batch should be compressed. A single table that requires
+ *                           temporary and output memory above this limit is allowed but will
+ *                           be compressed individually.
+ */
+abstract class BatchedTableCompressor(maxBatchMemorySize: Long) extends AutoCloseable with Arm {
+  private[this] val tables = new ArrayBuffer[ContiguousTable]
+  private[this] val tempBuffers = new ArrayBuffer[DeviceMemoryBuffer]
+  private[this] val oversizedOutBuffers = new ArrayBuffer[DeviceMemoryBuffer]
+  private[this] val results = new ArrayBuffer[CompressedTable]
+
+  // temporary and output memory being used as part of the current batch
+  private[this] var batchMemUsed: Long = 0
+
+  /**
+   * Add a contiguous table to be batch-compressed. Ownership of the table is transferred to the
+   * batch compressor which is responsible for closing the table.
+   * @param contigTable the contiguous table to be compressed
+   */
+  def addTable(contigTable: ContiguousTable): Unit = {
```

Comment:
Yes, the batch compressor itself must be closed at some point.  Note that `addTable` takes ownership of the table up to the point where it is placed into the `tables` tracking array in the batch compressor.  At that point `addTable` is no longer responsible for it (hence the `closeOnExcept` clause used in that method instead of `withResource`), and the table is now owned by the batch compressor.  Once the batch compressor is closed, all tables in the `tables` array will be closed.